feat: add BigIntFromString codec

EricCrosson opened the pull request:
Note:
- empty strings decode into a `left`
- whitespace-only strings decode into a `left`

This mirrors the behavior in the NumberFromString codec.

I reference version 0.5.11 in the docstring but haven't bumped to this version yet.

EricCrosson commented on src/BigIntFromString.ts:
```diff
@@ -0,0 +1,41 @@
+ */
+export const BigIntFromString: BigIntFromStringC = new t.Type<bigint, string, unknown>(
+  'BigIntFromString',
+  // tslint:disable-next-line
```
I'm not fond of this complete-disabling of tslint on this line, but I couldn't get [this rule](https://palantir.github.io/tslint/rules/typeof-compare/) to stop complaining. I took inspiration from [this io-ts commit](https://github.com/gcanti/io-ts/pull/417/files#diff-f41e9d04a45c83f3b6f6e630f10117feR374). If anybody knows how to appease tslint here I'd appreciate the assist.

mlegenhausen commented on src/BigIntFromString.ts:
```diff
@@ -0,0 +1,41 @@
+export const BigIntFromString: BigIntFromStringC = new t.Type<bigint, string, unknown>(
+  'BigIntFromString',
+  // tslint:disable-next-line
+  (u): u is bigint => typeof u === 'bigint',
```
You can reuse `bigint.is` here.

EricCrosson replied:
This will necessitate bumping the peer dep of io-ts from 2.0.0 to 2.1.0 (according to the release notes, but that still didn't work on my end strangely) and I feel replicating that one line is worth avoiding a possible breaking change downstream

mlegenhausen replied:
I think the bump makes sense cause using an `io-ts-types` version with bigint support before `io-ts` had that support confuses more.

You need to modify the `tsconfig.json` to add `BigInt` support to this project: https://github.com/gcanti/io-ts/blob/master/tsconfig.json#L17

EricCrosson replied:
Ok, I'll bump the `io-ts` version then. Would that be a minor or patch version bump for `io-ts-types`? I'll have to update the docstrings to match (still at 0.5.11 expecting a patch bump)

I found the root cause of my issue with the 2.0.0 -> 2.1.0 bump earlier: I only bumped the peerDependency section and not the devDependency. Once I bump and install io-ts 2.1.0, my tests pass without the `ESNext.BigInt` lib. But I don't know if it's wise to leave it out, so I added it

mlegenhausen replied:
I would also expect a patch bump cause the `io-ts` bump was just minor.

EricCrosson replied:
Sounds good. Thanks for your review @mlegenhausen, please let me know if there's anything else I can help with before merging

mlegenhausen replied:
Just wait for the final review from @gcanti

gcanti replied:
> I feel replicating that one line is worth avoiding a possible breaking change downstream

I agree with @EricCrosson here

EricCrosson replied:
Thanks for your input @gcanti. I reverted the latest changes to the type guard and all edits to `package.json` and `package-lock.json`

mlegenhausen commented on src/BigIntFromString.ts:
```diff
@@ -0,0 +1,41 @@
+    either.chain(t.string.validate(u, c), s => {
+      if (!NonEmptyString.is(s.trim())) {
+        return t.failure(u, c)
+      }
```
Don't `trim`. Codecs should make as few assumptions as possible. Trimming could be added by another Codec when needed.

EricCrosson replied:
The trim is necessary because `BigInt('') === BigInt(0)` -- we want to avoid that to keep parity with NumberFromString here https://github.com/gcanti/io-ts-types/blob/master/src/NumberFromString.ts#L27 -- but `BigInt(' ') === BigInt(0)` as well. Seems like a trap to decode `BigInt('')` into a left but `BigInt(' ')` into a right

mlegenhausen replied:
I see, after 10 years still learning new oddities in JS type conversion 😉

EricCrosson replied:
haha you and me both!

EricCrosson committed:
feat: add BigIntFromString codec

Note:
- empty strings decode into a `left`
- whitespace-only strings decode into a `left`

gcanti commented:
Thanks @EricCrosson @mlegenhausen
